EOS-19807:Cortx-1.0 - IPMITool | More monitoring for faulty resources.

satish-seagate opened the pull request:
Signed-off-by: Satish Darade <satish.darade@seagate.com>

## Problem Statement
<pre>
  <code>
    Story Ref (if any):
    Please add Problem statement here...
  </code>
</pre>
## Problem Description
<pre>
  <code>
    Please add Problem decription here...
  </code>
</pre>
## Solution
<pre>
  <code>
    Please add summary of the change,List any dependencies that are required for this change.
  </code>
</pre>
## Sanity testing on RPM done
<pre>
  <code>
    - [ ] Yes
    - [ ] No
  </code>
</pre>
## Unit/Manual Testing Description
<pre>
  <code>
    Please describe the tests that you ran to verify your changes.
  </code>
</pre>

mariyappanp commented on low-level/sensors/impl/generic/node_hw.py:
```diff
@@ -480,6 +481,54 @@ def run(self):
 
+    def _check_missmatch_resource_status(self):
+        for sensor_id in self.faulty_resources:
+            if self.faulty_resources[sensor_id]['resource'] in ['fan', 'disk', 'temperature', 'voltage']:
```
We need to include 'Power Supply' fru as well here.
Below would be suffice. Please check.
`if self.faulty_resources[sensor_id]['resource'].title() in self.fru_types.keys():`

mariyappanp commented on low-level/sensors/impl/generic/node_hw.py:
```diff
@@ -480,6 +481,54 @@ def run(self):
+                        self._create_sel_event_entry(sensor_id)
+
+    def _create_sel_event_entry(self, sensor_id):
+        event_map_dict = {
+            'Lower Non-Recoverable': 'lnr',
+            'Lower Critical': 'lcr',
+            'Lower Non-Critical': 'lnc',
+            'Upper Non-Critical': 'unc',
+            'Upper Critical': 'ucr',
+            'Upper Non-Recoverable': 'unr'
+        }
```
These events are specific to 'Fan' only.
How other fru events are handled?

satish-seagate replied:
We have possible mapping events for FAN, Temperature, voltage.
[root@sm17-r18 satish]# ipmitool event FAN1
Finding sensor FAN1... ok
Sensor States:
  lnr : Lower Non-Recoverable
  lcr : Lower Critical
  lnc : Lower Non-Critical
  unc : Upper Non-Critical
  ucr : Upper Critical
  unr : Upper Non-Recoverable
  
and for resources no mapping needed.
[root@sm17-r18 satish]# ipmitool event 'PS1 Status'
Finding sensor PS1 Status... ok
Sensor States:
  Presence detected
  Failure detected
  Predictive failure
  Power Supply AC lost
  AC lost or out-of-range
  AC out-of-range, but present
  Config Error
  Power Supply Inactive

mariyappanp commented on low-level/sensors/impl/generic/node_hw.py:
```diff
@@ -480,6 +481,54 @@ def run(self):
+            'Upper Non-Recoverable': 'unr'
+        }
+        event = self.faulty_resources[sensor_id]['event']
+        if "going high" in event.lower() or "going low" in event.lower():
```
Simplify as
`if event.lower() in ["going high", "going low" ]:`

mariyappanp commented on low-level/sensors/impl/generic/node_hw.py:
```diff
@@ -480,6 +481,54 @@ def run(self):
+        status = 'deasserted' if status == 'asserted' else 'asserted'
+        logger.warn(f"Found missing entry in the sel list for %s resource, \
+            Updating the sel list manually" % (sensor_id))
+        sel_event, retcode = self._run_ipmitool_subcommand("event '%s' '%s' %s" % (
```
Ensure Fan events are running for fan sensor and PSU events are running for PSU sensor and also Disk/Drive with its respective event.

satish-seagate replied:
Yes. That would be taken care by "self.faulty_resources" varibale.

mariyappanp commented on low-level/sensors/impl/generic/node_hw.py:
```diff
@@ -1016,6 +1084,15 @@ def _parse_psu_supply_info(self, index, date, _time, sensor, sensor_num, event,
+            self.faulty_resources[sensor_id] = {
+                'status': status,
+                'event': event,
+                'resource': 'psu_supply'
```
Resource name is already defined one. Please match with this.
```
    TYPE_PSU_SUPPLY = 'Power Supply'
```

satish-seagate replied:
Done

mariyappanp commented on low-level/sensors/impl/generic/node_hw.py:
```diff
@@ -1092,6 +1171,15 @@ def _parse_psu_unit_info(self, index, date, _time, sensor, sensor_num, event, st
+            self.faulty_resources[sensor_id] = {
+                'status': status,
+                'event': event,
+                'resource': 'psu_unit'
```
Resource name is already defined one. Please match with this.
```
    TYPE_PSU_UNIT = 'Power Unit'
```

satish-seagate replied:
Done

mariyappanp commented on low-level/sensors/impl/generic/node_hw.py:
```diff
@@ -958,6 +1013,15 @@ def _parse_fan_info(self, index, date, _time, device_id, sensor_id,\
+            self.faulty_resources[sensor_name] = {
+                'status': original_event,
+                'event': original_status,
+                'resource': 'fan'
```
Resource value should be 'Fan' instead of 'fan'.
Resource name is already defined one. Please match with this.
```
    TYPE_FAN = 'Fan'
```

satish-seagate replied:
Done

mariyappanp commented on low-level/sensors/impl/generic/node_hw.py:
```diff
@@ -1228,6 +1334,15 @@ def _parse_voltage_info(self, index, date, _time, sensor, sensor_num, event, sta
+            self.faulty_resources[sensor_name] = {
+                'status': status,
+                'event': event,
+                'resource': 'voltage'
```
This should be corrected as mentioned for other fru types.
Please take care.

satish-seagate replied:
Done

mariyappanp commented on low-level/sensors/impl/generic/node_hw.py:
```diff
@@ -1145,6 +1233,15 @@ def _parse_disk_info(self, index, date, _time, sensor, sensor_num, event, status
+                self.faulty_resources[sensor_id] = {
+                    'status': status,
+                    'event': event,
+                    'resource': 'disk'
```
This should be corrected as mentioned for other fru types.
Please take care.

satish-seagate replied:
Done

mariyappanp commented on low-level/sensors/impl/generic/node_hw.py:
```diff
@@ -1187,6 +1284,15 @@ def _parse_temperature_info(self, index, date, _time, sensor, sensor_num, event,
+            self.faulty_resources[sensor_name] = {
+                'status': status,
+                'event': event,
+                'resource': 'temperature'
```
This should be corrected as mentioned for other fru types.
Please take care.

satish-seagate replied:
Done

cdeshmukh commented on low-level/sensors/impl/generic/node_hw.py:
```diff
@@ -480,6 +481,58 @@ def run(self):
+        logger.warn("Found missing entry in the ipmitool sel list for %s resource,\
+            Updating the sel list manually" % (sensor_id))
+        # ipmitool event 'PS1 Status' 'Presence detected' deasserted
+        event_cmd = "event '%s' '%s' %s" % (sensor_id, event, status)
```
event command is to trigger a test event, not supposed to be used to for any user generated event, check its man page

>event
\<predefined event number N\>
Send a pre-defined test event to the System Event Log. The following events are included as a means to test the functionality of the System Event Log component of the BMC (an entry will be added each time the event N command is executed).

>Currently supported values for N are:
1 Temperature: Upper Critical: Going High
2 Voltage Threshold: Lower Critical: Going Low
3 Memory: Correctable ECC

>NOTE: These pre-defined events will likely not produce "accurate" SEL records for a particular system because they will not be correctly tied to a valid sensor number, but they are sufficient to verify correct operation of the SEL.

satish-seagate replied:
Done

cdeshmukh commented on low-level/sensors/impl/generic/node_hw.py:
```diff
@@ -480,6 +481,58 @@ def run(self):
+                            and sensor_get_data['Status'] == 'ok':
+                            self._create_sel_event_entry(sensor_id)
+
+    def _create_sel_event_entry(self, sensor_id):
```
Dont need this function, we are not going to introduce any SEL entry ourselves

satish-seagate replied:
Yes. deleted.
Done

cdeshmukh commented on low-level/sensors/impl/generic/node_hw.py:
```diff
@@ -554,6 +607,8 @@ def _notify_NodeDataMsgHandler(self):
 
         if last_index is not None:
             self._write_index_file(last_index)
+        else:
```
Doesnt need to be under 'else' case

After SEL processing is done, check for faulty resources explicitly. What was faulty resource last fault recorded, and get sensor/sdr directly and see if its rectified, if yes, raise required alert & remove from faulty alerts list

satish-seagate replied:
Done

satish-seagate commented on low-level/sensors/impl/generic/node_hw.py:
```diff
@@ -480,6 +481,58 @@ def run(self):
             logger.warning(f"{self.SENSOR_NAME} Node hw monitoring disabled")
             self.shutdown()
 
+    def _check_missmatch_resource_status(self):
```
added description and changed function name.
Done

satish-seagate commented on low-level/sensors/impl/generic/node_hw.py:
```diff
@@ -480,6 +481,58 @@ def run(self):
+            if self.faulty_resources[sensor_id]['resource'] in self.fru_types.keys():
+                if self.faulty_resources[sensor_id]['resource'] in [
+                    self.TYPE_PSU_SUPPLY, self.TYPE_PSU_UNIT]:
+                    sdr_get_data, _ = self._get_sensor_sdr_props(sensor_id)
```
Done.

satish-seagate commented on low-level/sensors/impl/generic/node_hw.py:
```diff
@@ -480,6 +481,58 @@ def run(self):
+        for sensor_id in self.faulty_resources:
+            if self.faulty_resources[sensor_id]['resource'] in self.fru_types.keys():
+                if self.faulty_resources[sensor_id]['resource'] in [
+                    self.TYPE_PSU_SUPPLY, self.TYPE_PSU_UNIT]:
```
Removed that check. Please check.
Done

satish-seagate commented on low-level/sensors/impl/generic/node_hw.py:
```diff
@@ -480,6 +481,58 @@ def run(self):
+                        # 'Assertion Events': ''
+                        if sensor_get_data['Assertion Events'] in [""] \
+                            and sensor_get_data['Status'] == 'ok':
+                            self._create_sel_event_entry(sensor_id)
```
Done

cdeshmukh commented on low-level/sensors/impl/generic/node_hw.py:
```diff
@@ -554,6 +612,7 @@ def _notify_NodeDataMsgHandler(self):
 
         if last_index is not None:
             self._write_index_file(last_index)
+        self._check_faulty_resource_status()
```
move _check_faulty_resource_status() to line 472 above, as we want to keep it separate from SEL processing and that too execute only after clean SEL processing, which is confirmed after second call to _notify_NodeDataMsgHandler() on line 471

satish-seagate replied:
Yes agreed, Done.
Addressed this review comments in https://github.com/Seagate/cortx-monitor/pull/393

cdeshmukh commented on low-level/sensors/impl/generic/node_hw.py:
```diff
@@ -480,6 +481,63 @@ def run(self):
+
+    def _generate_resource_alert(self, sensor_id, faulty_res):
+        status = faulty_res[sensor_id]['status'].lower()
+        status = 'Deasserted' if status == 'asserted' else 'Asserted'
```
>status = 'Deasserted' if status == 'asserted'

Why status is inverted here,  to 'Deasserted' or vice versa?

satish-seagate replied:
Any fault resolved resources status is inverted of faulty resources or vice versa.
Example: (Voltage)
**8 | 05/17/2021 | 19:23:31 | Power Supply #0xc4 | Failure detected () | Asserted**
9 | 05/17/2021 | 19:23:32 | Fan #0x45 | Lower Critical going low  | Asserted
**a | 05/17/2021 | 19:29:25 | Power Supply #0xc4 | Failure detected () | Deasserted**

Addressed this review comments in https://github.com/Seagate/cortx-monitor/pull/393

cdeshmukh commented on low-level/sensors/impl/generic/node_hw.py:
```diff
@@ -480,6 +481,63 @@ def run(self):
+        is_last = True
+        index, date, event_time, device_id, device_type, sensor_num, event, \
+            status = self._make_sel_event(sel_line)
+        logger.warn("Found missing entry in the ipmitool sel list for %s resource."
```
> Found missing entry in the IPMI SEL for resource "%s" & event "%s"

satish-seagate replied:
Done.
Addressed this review comments in https://github.com/Seagate/cortx-monitor/pull/393

cdeshmukh commented on low-level/sensors/impl/generic/node_hw.py:
```diff
@@ -480,6 +481,63 @@ def run(self):
+        sel_line = f"0 | {date_time} | {device_id} | {event} | {status}"
+        is_last = True
+        index, date, event_time, device_id, device_type, sensor_num, event, \
+            status = self._make_sel_event(sel_line)
```
device_type, sensor_num are the only one we dont have handy here, why to feed whole details to _make_sel_event & then get again, extract it the way its done in make_sel_event.

satish-seagate replied:
Created separate functions and calling that function from both places. Done.
Addressed this review comments in https://github.com/Seagate/cortx-monitor/pull/393

cdeshmukh commented on low-level/sensors/impl/generic/node_hw.py:
```diff
@@ -480,6 +481,63 @@ def run(self):
+        # here we are polling resources that are detected to be faulty or missing
+        # periodically, in addition to current polling for SEL.
+
+        # Copying object to avoid RuntimeError: dictionary changed size during iteration
```
need to have it under try - exception block, there are chances of exception here, include _generate_resource_alert() call

satish-seagate replied:
Done.
Addressed this review comments in https://github.com/Seagate/cortx-monitor/pull/393

cdeshmukh commented:
@satish-seagate please close this PR as we decided to avoid any checkins to cortx-1.0 until absolutely necessary & marked as must by Ujjwal / Amit/ Gregory T / Support Team

satish-seagate commented on low-level/sensors/impl/generic/node_hw.py:
```diff
@@ -480,6 +481,63 @@ def run(self):
+        # resources fault & recovery. However, its been observed multiple times
+        # with Supermicro servers, that SEL not getting updated in case recoveries
+        # of some resources.
+        # here we are polling resources that are detected to be faulty or missing
```
Done.
Addressed this review comments in https://github.com/Seagate/cortx-monitor/pull/393

satish-seagate commented on low-level/sensors/impl/generic/node_hw.py:
```diff
@@ -1228,6 +1351,18 @@ def _parse_voltage_info(self, index, date, _time, sensor, sensor_num, event, sta
             "description": event
         }
 
+        if (threshold.lower() in ['low', 'high'] and status.lower() == "asserted"):
```
Done.
Addressed this review comments in https://github.com/Seagate/cortx-monitor/pull/393

satish-seagate commented on low-level/sensors/impl/generic/node_hw.py:
```diff
@@ -178,6 +178,7 @@ def __init__(self):
             # to test the current sensor
             # self.TYPE_CURRENT: self._parse_current_info,
         }
+        self.faulty_resources = {}
```
Done.
Addressed this review comments in https://github.com/Seagate/cortx-monitor/pull/393

satish-seagate commented:
I am closing this PR as we decided to avoid any check-ins to cortx-1.0 until absolutely necessary & marked as a must by @ujjwalpl  / @amitwac2608 / @gregnsk / Support Team.
https://github.com/Seagate/cortx-monitor/pull/390#issuecomment-842332535

@cdeshmukh I have addressed all comments from this PR to https://github.com/Seagate/cortx-monitor/pull/393. Please review
